src/ORM/ArrayList.php

```diff
@@ -45,13 +51,12 @@ class ArrayList extends ViewableData implements SS_List, Filterable, Sortable, L
     /**
      * Holds the items in the list
      *
-     * @var array
+     * @var array<array-key, T>
      */
     protected $items = [];
 
     /**
-     *
-     * @param array $items - an initial array to fill this object with
+     * @param array<array-key, T> $items - an initial array to fill this object with
```

Comment:
From the description:

- Because ArrayList can accept values of mixed types (you could have a combination of ArrayData, DataObject, and some custom class for example), the return values might not always be reported correctly, especially if adding objects to the list after instantiation. This is acceptable since that sort of usage is less common - and they'd already have to use @var or similar in their IDE before this PR.

What actually sets the `@template T` when you pass in a bunch of mixed type arguments in the constructor? The first value? The last value? (it changes on each item added)? Or do you need to call `setDataClass()` directly?

> This is acceptable since that sort of usage is less common - and they'd already have to use @var or similar in their IDE before this PR.

> Ideally people should be wrapping their associative arrays in ArrayData anyway because of the existing inconsistency with what ArrayList returns.

That's debatable that it's acceptable? I think we're better off just not trying to force using generics in ArrayList if we really should just be using mixed here instead, rather that saying it's the users fault for doing it wrong (despite it being perfectly allowable)

Reply:
> What actually sets the `@template T` when you pass in a bunch of mixed type arguments in the constructor? The first value? The last value? (it changes on each item added)? Or do you need to call `setDataClass()` directly?

I'm not sure. That's an edge case that I'm not sure we need to cater for - as I mentioned in the description, to get the correct type when your list contains mixed types you already have to use `@var`, and that doesn't change with this PR.

This PR gives you correct types when you use only a specific type - and from what I can tell it _doesn't_ deteriorate the DX when using mixed types.

With that in mind, do you think we should revert this change or leave it in? I can go either way.

Reply:
> And besides all that, most of the time when instantiating a new arraylist you'll use ArrayList::create(), so your IDE won't know what's being passed to the constructor anyway, so you'll see T as mixed.

How come this doesn't work?  I tested this locally and these two behave differently:

```php
// Type hints MyDataObject
$list = new ArrayList([
    new MyDataObject()
]);
// Won't type hint MyDataObject
$list = ArrayList::create([
    new MyDataObject()
]);
foreach($list as $obj) {
    $obj->mymethod; // defined on MyDataObject, only new ArrayList() will type hint it
}

> I suspect this will be by far the more common scenario - having generics in Arraylist allows you to use @return ArrayList<SomeType>

By this do you mean project code will add this docblock return type to a method defined in the project?

Reply:
> > And besides all that, most of the time when instantiating a new arraylist you'll use ArrayList::create(), so your IDE won't know what's being passed to the constructor anyway, so you'll see T as mixed.
> How come this doesn't work?

`create()` isn't the constructor, so it doesn't have the generic typehints available. Static analysis tools could probably be made to understand what's going on there but IDEs aren't that smart.

> > I suspect this will be by far the more common scenario - having generics in Arraylist allows you to use `@return` ArrayList
> By this do you mean project code will add this docblock return type to a method defined in the project?

Projects _can_ add those, but I honestly don't expect them to. I mostly mean that _we_ will use the generic return types (e.g. `@return ArrayList<Member>`), and projects can then benefit from those by their IDEs correctly identifying the contents of the returned lists.